### PR TITLE
TA-712: Add watch Feature 

### PR DESCRIPTION
**Pull request by itouir**
Tasks are used by exoers to manage effeciency their work, but they didn't receive enough notifications for every modification done on their list of tasks. Here a Watch feature is added to the Task addon in order to get all related notifications.
What's already done :

- A watch icon is added near the already available actions in the top right of a task
- The watch icon isn't be displayed for the task's creator, the assignee and the coworker
- When the user clicks on the watch icon, it turns to unwatch status
- When a user clicks on the watch icon, a popup message appears "You start watching the task", user will receive all notificationsrelated to: title, assignee, coworker, due date, status, description, schedule, priority, When the task is marked as completed, When the task has been commented
- When a user clicks on the Unwatch icon, a popup message appears "You stop watching the task", no notification received.

Ps: [Video](https://drive.google.com/file/d/1bVCXWXyE93tk8xx5Hy5tQG4xZiVXCer_/view)

### CODE REVIEW

File: services/src/test/resources/locale/portlet/taskManagement_en.properties [hunk 1]
```diff
@@ -126,6 +127,8 @@ popup.msg.cloneProject=You are about to clone project <strong>{}</strong>. Pleas
 popup.msg.deleteTask=Are you sure you want to delete this task : '{}' ?
 popup.checkbox.cloneTask=also clone uncompleted tasks
 popup.checkbox.project.deleteChild=also delete all sub-projects
+popup.watch.task=You start watching the task.
```
Comment (ahamdi) [inline services/src/test/resources/locale/portlet/taskManagement_en.properties:130]:
```suggestion
popup.watch.task=You started watching the task.
```

File: services/src/test/resources/locale/portlet/taskManagement_en.properties [hunk 2]
```diff
@@ -126,6 +127,8 @@ popup.msg.cloneProject=You are about to clone project <strong>{}</strong>. Pleas
...
+popup.unwatch.task=You stop watching the task.
```
Comment (ahamdi) [inline services/src/test/resources/locale/portlet/taskManagement_en.properties:131]:
```suggestion
popup.unwatch.task=You stopped watching the task.
```

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java [hunk 3]
```diff
@@ -239,7 +243,30 @@ public Response clone(Long id) throws EntityNotFoundException, JSONException, Un
...
-
+  @Resource
+  @Ajax
+  @MimeType.HTML
+  public Response watchTask(Long id) throws Exception, EntityNotFoundException {
+    Task task = taskService.getTask(id);
+     Boolean isWatched = detectWatched(true,task);
+     if (!isWatched) {
+        //String msg= bundle.getString("popup.watch.task");
+       String msg="You start watching the task.";
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java:254]: Messages should be i18n

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java [hunk 4]
```diff
@@ -239,7 +243,30 @@ public Response clone(Long id) throws EntityNotFoundException, JSONException, Un
       json.put("id", newTask.getId()); //Can throw JSONException
       return Response.ok(json.toString());
   }
-
+  @Resource
+  @Ajax
+  @MimeType.HTML
+  public Response watchTask(Long id) throws Exception, EntityNotFoundException {
+    Task task = taskService.getTask(id);
+     Boolean isWatched = detectWatched(true,task);
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java:251]: This should be inside the TaskService

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java [hunk 5]
```diff
@@ -1050,4 +1077,23 @@ private CommentModel addCommentModel(Comment cmt, List<CommentModel> listComment
     listComments.add(commentModel);
     return commentModel;
   }
+  public static boolean detectWatched(boolean state, Task task) throws Exception {
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java:1080]: Please remove this function from the controller (We should not add logic in controller) and place its logic inside the TaskService

File: services/src/main/java/org/exoplatform/task/util/TaskUtil.java [hunk 6]
```diff
@@ -1079,5 +1090,19 @@ private static TaskService getTaskService() {
     return getTaskService().getCoworker(taskId);
   }
 
+  public static Set<String> getWatcher(TaskService taskService, Task task) {
+    return taskService.getWatchersOfTask(task);
+  }
+  public static boolean isEligibleWatcher(Task task){
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/task/util/TaskUtil.java:1096]: This is a static call, it should get the username in a parameter and don't rely on the actual context.

File: integration/src/main/java/org/exoplatform/task/integration/notification/MailTemplateProvider.java [hunk 7]
```diff
@@ -61,7 +61,13 @@
...
     @TemplateConfig(pluginId = TaskCompletedPlugin.ID, template = "war:/notification/templates/mail/TaskCompletedPlugin.gtmpl"),
     @TemplateConfig(pluginId = TaskCommentPlugin.ID, template = "war:/notification/templates/mail/TaskCommentPlugin.gtmpl"),
-    @TemplateConfig(pluginId = TaskMentionPlugin.ID, template = "war:/notification/templates/mail/TaskMentionPlugin.gtmpl") })
+    @TemplateConfig(pluginId = TaskMentionPlugin.ID, template = "war:/notification/templates/mail/TaskMentionPlugin.gtmpl"),
+    @TemplateConfig(pluginId = TaskTitlePlugin.ID, template = "war:/notification/templates/mail/TaskTitlePlugin.gtmpl"),
+    @TemplateConfig(pluginId = TaskDescriptionPlugin.ID, template = "war:/notification/templates/mail/TaskDescriptionPlugin.gtmpl"),
+    @TemplateConfig(pluginId = TaskStatusPlugin.ID, template = "war:/notification/templates/mail/TaskStatusPlugin.gtmpl"),
+    @TemplateConfig(pluginId = TaskSchedulePlugin.ID, template = "war:/notification/templates/mail/TaskSchedulePlugin.gtmpl"),
+    @TemplateConfig(pluginId = TaskPriorityPlugin.ID, template = "war:/notification/templates/mail/TaskPriorityPlugin.gtmpl") })
```
Comment (ahamdi) [inline integration/src/main/java/org/exoplatform/task/integration/notification/MailTemplateProvider.java:69]: Please put all those plugins inside one plugin, TaskEditionPlugin for example

File: services/src/main/java/org/exoplatform/task/util/TaskUtil.java [hunk 8]
```diff
@@ -1079,5 +1090,19 @@ private static TaskService getTaskService() {
...
+    Identity identity = ConversationState.getCurrent().getIdentity();
+    String username = identity.getUserId();
+    Set <String> coworker=task.getCoworker();
+    String assignee=task.getAssignee();
+    String owner=task.getCreatedBy();
```
[the remark this answers is not in the document]
Comment (itouir) [reply services/src/main/java/org/exoplatform/task/util/TaskUtil.java]: Code updated

File: services/src/main/resources/db/changelog/task.db.changelog-1.0.0.xml [hunk 9]
```diff
@@ -370,6 +370,21 @@
     </modifySql>    
   </changeSet>
 
+  <!-- Definition of TASK_TASK_WATCHERS table -->
```
[the remark this answers is not in the document]
Comment (itouir) [reply services/src/main/resources/db/changelog/task.db.changelog-1.0.0.xml]: done

File: services/src/main/resources/db/changelog/task.db.changelog-1.0.0.xml [hunk 10]
```diff
@@ -396,6 +411,10 @@
                  tableName="TASK_TASKS">
       <column name="CREATED_BY" type="NVARCHAR(50)"/>
     </createIndex>
+    <createIndex indexName="IDX_TASK_TASK_WATCHERS_01"
```
[the remark this answers is not in the document]
Comment (itouir) [reply services/src/main/resources/db/changelog/task.db.changelog-1.0.0.xml]: done

File: services/src/test/resources/locale/portlet/taskManagement_en.properties (hunk 1 above)
Comment (itouir) [reply services/src/test/resources/locale/portlet/taskManagement_en.properties]: done

File: services/src/test/resources/locale/portlet/taskManagement_en.properties (hunk 2 above)
Comment (itouir) [reply services/src/test/resources/locale/portlet/taskManagement_en.properties]: done

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java (hunk 3 above)
Comment (itouir) [reply task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java]: Yes already fixed

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java (hunk 5 above)
Comment (itouir) [reply task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java]: done

File: services/src/main/java/org/exoplatform/task/util/TaskUtil.java (hunk 6 above)
Comment (itouir) [reply services/src/main/java/org/exoplatform/task/util/TaskUtil.java]: done

File: integration/src/main/java/org/exoplatform/task/integration/notification/MailTemplateProvider.java (hunk 7 above)
Comment (itouir) [reply integration/src/main/java/org/exoplatform/task/integration/notification/MailTemplateProvider.java]: done

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java [hunk 11]
```diff
@@ -239,7 +244,31 @@ public Response clone(Long id) throws EntityNotFoundException, JSONException, Un
       json.put("id", newTask.getId()); //Can throw JSONException
       return Response.ok(json.toString());
   }
+  @Resource
+  @Ajax
+  @MimeType.HTML
+  public Response OpenPopupWatch(Long id) throws Exception, EntityNotFoundException {
+    Task task = taskService.getTask(id);
+     Boolean isWatched = taskService.detectWatched(true,task);
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java:252]: Why should you check database to know if the task is watched or not ? couldn't we get the information directly from the already loaded Task object ?

File: services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java [hunk 12]
```diff
@@ -209,6 +211,26 @@ public ChangeLog addTaskLog(long id, String username, String actionName, String
     return daoHandler.getTaskLogHandler().create(log);
   }
 
+  @Override
+  public  boolean detectWatched(boolean state, Task task) throws Exception {
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java:215]: You should pass the username as a parameter, don't use ConversationState.getCurrent() isn services layer

File: services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java [hunk 13]
```diff
@@ -209,6 +211,26 @@ public ChangeLog addTaskLog(long id, String username, String actionName, String
...
+    ConversationState conversationState = ConversationState.getCurrent();
+    String currentUserId = conversationState.getIdentity().getUserId();
+    boolean isWatched=false;
+    Set<String> watchers = getWatchersOfTask(task);
+    for (String watcher : watchers) {
+      if (watcher.equals(currentUserId))
+        isWatched = true;
+    }
+    if (state) {
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java:224]: Why does state refer to ?

File: task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl [hunk 14]
```diff
@@ -1,9 +1,15 @@
...
 #{param name=bundle/}
+#{param name=taskService/}
 #{param name=useCalendar/}
 
 <%
+import org.exoplatform.services.security.Identity
+import org.exoplatform.task.management.controller.TaskController
+import org.exoplatform.task.util.ProjectUtil
+import org.exoplatform.services.security.ConversationState
+
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl:12]: Unused imports, please check if you still need them

File: task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl [hunk 15]
```diff
@@ -26,19 +32,29 @@ def commentCount = taskModel.getCommentCount();
 def comments = taskModel.getComments();
 def currentUser = taskModel.getCurrentUser();
 def labels = taskModel.getLabels();
+boolean isWatched= taskService.detectWatched(false,task);
```
Comment (ahamdi) [inline task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl:35]: taskService.detectWatched should just take the task and the username and returns true if the task is watched.

File: services/src/main/java/org/exoplatform/task/util/TaskUtil.java [hunk 16]
```diff
@@ -1079,5 +1090,14 @@ private static TaskService getTaskService() {
     return getTaskService().getCoworker(taskId);
   }
 
+  public static Set<String> getWatcher(TaskService taskService, Task task) {
+    return taskService.getWatchersOfTask(task);
+  }
+  public static boolean isEligibleWatcher(Task task,String username){
+    if(!task.getCoworker().contains(username)&& !username.equals(task.getAssignee())&& !username.equals(task.getCreatedBy()))
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/task/util/TaskUtil.java:1097]: Format code, use brackets for the if/else block and respect indentation

File: services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java (hunk 13 above)
Comment (itouir) [reply services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java:224]: To Control if an action is done or not (the switch off/on)

File: services/src/main/java/org/exoplatform/task/util/TaskUtil.java (hunk 16 above)
Comment (itouir) [reply services/src/main/java/org/exoplatform/task/util/TaskUtil.java]: done

File: task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl (hunk 15 above)
Comment (itouir) [reply task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl]: updated

File: task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl (hunk 14 above)
Comment (itouir) [reply task-management/src/main/java/org/exoplatform/task/management/templates/detail.gtmpl]: done

File: services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java (hunk 12 above)
Comment (itouir) [reply services/src/main/java/org/exoplatform/task/service/impl/TaskServiceImpl.java]: done

File: task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java (hunk 11 above)
Comment (itouir) [reply task-management/src/main/java/org/exoplatform/task/management/controller/TaskController.java]: updated

File: services/src/test/java/org/exoplatform/task/dao/TestTaskDAO.java [hunk 17]
```diff
@@ -702,7 +702,58 @@ public void testGetOverdueTasks() {
...
-  
+
+  @Test
+  public void testAddWatchersToTasks() throws Exception {
+    // Test Add watchers to task
+    String username="john";
+    Task task = newTaskInstance("task ", "description of task ", username);
+    tDAO.create(task);
+    Set<String> watchers = new HashSet<>();
+    tDAO.addWatcherToTask(username,task);
```
Comment (ahamdi) [inline services/src/test/java/org/exoplatform/task/dao/TestTaskDAO.java:713]: use another username for watchers other than the assignee.

Comment (itouir) [reply services/src/test/java/org/exoplatform/task/dao/TestTaskDAO.java]: done.

**Pull request merged**